core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java

```diff
@@ -35,6 +35,80 @@
+            new CrossReferenceBuilder<CitationDatabase>()
+                    .database(CitationDatabase.DOI)
+                    .id("doi-id3")
+                    .build();
+
+    @Test
+    void checkCitationIdIsPubMed() {
+        String title = "a title";
+        String publicationDate = "2015-MAY";
+        List<String> authoringGroup = asList("G1", "G2");
+        List<String> authors = asList("Tom", "John");
+        TestableCitation citation =
+                new TestableCitationBuilder()
+                        .title(title)
+                        .publicationDate(publicationDate)
+                        .authoringGroupsSet(authoringGroup)
+                        .authorsSet(authors)
+                        .citationCrossReferencesAdd(XREF1)
+                        .build();
+        assertThat(citation.getId(), is("id1"));
+    }
+
+    @Test
+    void checkCitationIdIsAgricola() {
+        String title = "a title";
+        String publicationDate = "2015-MAY";
+        List<String> authoringGroup = asList("G1", "G2");
+        List<String> authors = asList("Tom", "John");
+        TestableCitation citation =
+                new TestableCitationBuilder()
+                        .title(title)
+                        .publicationDate(publicationDate)
+                        .authoringGroupsSet(authoringGroup)
+                        .authorsSet(authors)
+                        .citationCrossReferencesAdd(XREF2)
+                        .build();
+        assertThat(citation.getId(), is("id2"));
+    }
+
+    @Test
+    void checkCitationIdIsDoiHashGenerated() {
+        String title = "a title";
+        String publicationDate = "2015-MAY";
+        List<String> authoringGroup = asList("G1", "G2");
+        List<String> authors = asList("Tom", "John");
+        TestableCitation citation =
+                new TestableCitationBuilder()
+                        .title(title)
+                        .publicationDate(publicationDate)
+                        .authoringGroupsSet(authoringGroup)
+                        .authorsSet(authors)
+                        .citationCrossReferencesAdd(XREF3)
+                        .build();
+        assertThat(citation.getId(), is("CI-2V34M1IJ67J60"));
+        assertThat(citation.generateHash("doi-id3"), is("CI-2V34M1IJ67J60"));
+    }
+
+    @Test
+    void checkCitationIdIsHashGenerated() {
+        String title = "a title";
+        String publicationDate = "2015-MAY";
+        List<String> authoringGroup = asList("G1", "G2");
+        List<String> authors = asList("Tom", "John");
+        TestableCitation citation =
+                new TestableCitationBuilder()
+                        .title(title)
+                        .publicationDate(publicationDate)
+                        .authoringGroupsSet(authoringGroup)
+                        .authorsSet(authors)
+                        .build();
+        assertThat(citation.getId(), is("CI-7TL5R4DGOC4O8"));
+    }
+
```

Comment:
I think, we should add a test when citationType is null

Reply:
I think Citation Type can not be null. See BookImpl constructor for example. We have the citationType in the constructor and it happens with all concrete class under citation interface. So, this is an use case that is impossible to replicate. Does it make sense to you?

Reply:
Yes, it will work with existing impl but if we need new impl and pass null, it will fail. 
We don't have validation, that is why i thought to have a defensive code